src/portable/synopsys/dwc2/dcd_dwc2.c

```diff
@@ -989,8 +1011,10 @@ void dcd_int_handler(uint8_t rhport) {
 
   if (gintsts & GINTSTS_USBRST) {
     // USBRST is start of reset.
+    DCD_ENTER_CRITICAL();
```

Comment:
this is only relevant on SMP mcu like P4 right ? since when app enter critical, this isr shoud not be running ?

Reply:
I think the same race condition can happen on single core MCUs too.
For example
1. Code from main app is modifying `xfer_ctl_t` and is interrupted
2. Content of `xfer_ctl_t` is now inconsistent
3. The ISR modifies `xfer_ctl_t` again
4. Main app continues

In steps 3 and 4 we can access inconsistent `xfer_ctl_t` which can lead to undefined behavior (division by zero in the cases I've seen)

Reply:
but when application modify xfer_ctl_t it already disable irq with enter_critical() the irq shouldn't be invoked. since usb irq is the only irq making change to this xfer ctl, it isn't be preempted by other irq. I guess it would be safe for signle core CPU ? I am trying to see if we could optional omit this for s2/s3 ?

Reply:
Yes, you're right. Although S3 is also dual-core, and the problem was reported and reproduced on S3 too.